### PR TITLE
DT-1361 Release a prisoner API

### PR DESCRIPTION
**Pull request by Mjwillis**

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java [hunk 1]
```diff
@@ -101,6 +105,21 @@ public InmateDetail getOffender(@Pattern(regexp = "^[A-Z]\\d{4}[A-Z]{2}$", messa
 
     @ApiResponses({
             @ApiResponse(code = 500, message = "Unrecoverable error occurred whilst processing request.", response = ErrorResponse.class)})
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java:107]: Do we need a 400 and 404 here as both are possible?

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/AgencyInternalLocation.java [hunk 2]
```diff
@@ -95,6 +99,14 @@ private boolean hasSpace() {
         return capacity != null && currentOccupancy != null && currentOccupancy < capacity;
     }
 
+    public Integer incrementCurrentOccupancy() {
+        if (currentOccupancy != null) {
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/AgencyInternalLocation.java:103]: Not sure how currentOccupancy could be null - but if so should we set it to 1 here?

File: src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerReleaseService.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    private final BedAssignmentHistoriesRepository bedAssignmentHistoriesRepository;
+    private final AgencyInternalLocationRepository agencyInternalLocationRepository;
+    private final MovementTypeAndReasonRespository movementTypeAndReasonRespository;
+    private final OffenderSentenceAdjustmentRepository offenderSentenceAdjustmentRepository;
+    private final OffenderKeyDateAdjustmentRepository offenderKeyDateAdjustmentRepository;
+
+    public void releasePrisoner(final String prisonerIdentifier, final String movementReasonCode, final String commentText) {
+
+        // check that prisoner is active in
+        final var optionalOffenderBooking = offenderBookingRepository.findByOffenderNomsIdAndBookingSequence(prisonerIdentifier, 1);
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerReleaseService.java:54]: Would booking sequence always be 1?

File: src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerReleaseService.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        if (!booking.getActiveFlag().equals("Y")) {
+            throw new BadRequestException("Prisoner is not currently active");
+        }
+
+        if (!booking.getInOutStatus().equals("IN")) {
+            throw new BadRequestException("Prisoner is not currently IN");
+        }
+
+        // unless you are system user you must have the caseload
+        if (!AuthenticationFacade.hasRoles("SYSTEM_USER")) {
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerReleaseService.java:67]: Can this be done with @VerifyBookingAccess (which accepts an override role I think)?

**Review by mikehalmamoj — APPROVED** [review]

File: src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerReleaseService.java (hunk 3 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerReleaseService.java:54]: yes

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/AgencyInternalLocation.java (hunk 2 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/AgencyInternalLocation.java:103]: agreed

File: src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerReleaseService.java (hunk 4 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerReleaseService.java:67]: Already done this bit in the resource class - removed

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**